### PR TITLE
m(breaking): Simplify internal storage and lookup mechanisms, take 2

### PR DESCRIPTION
**Pull request by notgull**
Rather than using `slab` like in #45, this PR just uses the ID as an index. This leaks memory when fonts are removed.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -88,15 +88,15 @@ pub use ttf_parser::Width as Stretch;
 /// as different strings, but does not make any guarantees about format or
 /// content of the strings.
 #[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd, Debug)]
-pub struct ID(u32);
+pub struct ID(usize);
```
Comment (RazrFalcon) [inline src/lib.rs:91]: We can keep it as `u32`. No one will have more that 4M font faces.

File: src/lib.rs [hunk 2]
```diff
@@ -450,12 +445,11 @@ impl Database {
     /// after loading a large directory with fonts.
     /// Or a specific face from a font.
     pub fn remove_face(&mut self, id: ID) -> bool {
```
Comment (RazrFalcon) [inline src/lib.rs:447]: I don't think we should return `bool` anymore.

And we should rename it to `disable_face`.

File: src/lib.rs [hunk 3]
```diff
@@ -714,6 +708,9 @@ pub struct FaceInfo {
 
     /// Indicates that the font face is monospaced.
     pub monospaced: bool,
+
+    /// Whether or not this font face has been disabled.
+    pub disabled: bool,
```
Comment (RazrFalcon) [inline src/lib.rs:713]: A disabled face should be ignored during querying.

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

**Comment by RazrFalcon** [comment]
While a simple `disabled` flag is fine, it will "leak" memory in case of `Source::Binary`.
Adding `Source::None` instead seems like a way more complicated change.

Will have to think about it.

I'm trying to think about a use case when someone would be constantly adding and removing fonts to a database. And I can think of one in case of SVG with embedded fonts, in which case removal of a font is necessary....

We need a 0(1) indexing, strong/unique indices and an ability to remove a face... Looks like a [generational arena](https://github.com/fitzgen/generational-arena) to me. It's better than `slab`, since there are no ABA problem anymore, but we also loose `&[T]` and our ID is 16 bytes instead of 4 now.

In the end,  `fontdb` is deceptively simple. It's actually surprisingly hard to have "just a list of fonts".

**Comment by notgull** [comment]
> but we also loose `&[T]` and our ID is 16 bytes instead of 4 now.

We could use `slotmap`, which is similar to `generational-arena` but uses an 8-byte key instead of a 16-byte one.

**Comment by RazrFalcon** [comment]
Yes, it seems like `slotmap` is our own solution. It will break the API a bit and would require MSRV bump as well.

**Comment by notgull** [comment]
Superseded by #47

**Pull request closed without merging**